### PR TITLE
Suggested spec change to resource locks (`lock`) interface for Terraform

### PR DESCRIPTION
**Pull request by MariusStorhaug**
# Overview/Summary

## This PR fixes/adds/changes/removes

- Updates the suggestion for the terraform interface `lock`
- Fixes: #423

### Breaking Changes

N/A

## As part of this Pull Request I have

- [x] Read the Contribution Guide and ensured this PR is compliant with the guide
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Azure-Verified-Modules/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/Azure-Verified-Modules/issues) or ADO Work Items (Internal Only)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Azure-Verified-Modules/)
- [x] Ensured PR tests are passing
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Docs etc.)

### CODE REVIEW

**Comment by MariusStorhaug** [comment]
@matt-FFFFFF: Opened a new PR with a working solution that should be closer to you suggestion while removing the "None" option.

**Comment by matt-FFFFFF** [comment]
Hi thanks for the PR

I don't like the mixed use of a default and optional.

I prefer this:

```terraform
    type = object({
      name = optional(string, null)
      kind = optional(string, null)
    })
   default  = {}
   nullable = false
```

But with this validation:

```terraform
    validation {
      condition     = var.lock.kind == null || try(contains(["CanNotDelete", "ReadOnly"], var.lock.kind), false)
      error_message = "The lock level must be one of: `null`, '\"CanNotDelete\"', or '\"ReadOnly\"'."
    }
```

Also note the use of escaped quotes to differentiate `null` and `"null"`

**Comment by MariusStorhaug** [comment]
I was thinking more that the `optional` keyword and function on `kind` would be 'lying' and not really showing the true spec of the object that is needed. Also when reading the docs it would indicate that `kind` is optional, while it is required if specifying the "lock" variable.

**Comment by matt-FFFFFF** [comment]
> I was thinking more that the `optional` keyword and function on `kind` would be 'lying' and not really showing the true spec of the object that is needed. Also when reading the docs it would indicate that `kind` is optional, while it is required if specifying the `lock` variable.

Disagree here. It's the same in the MI interface and is IMHO cleaner.

File: docs/static/includes/interfaces/int.locks.schema.tf [hunk 1]
```diff
@@ -1,21 +1,31 @@
 variable "lock" {
   type = object({
+    kind = string
```
Comment (matt-FFFFFF) [inline docs/static/includes/interfaces/int.locks.schema.tf:3]:
```suggestion
    kind = optional(string, null)
```

File: docs/static/includes/interfaces/int.locks.schema.tf [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
     name = optional(string, null)
-    kind = optional(string, "None")
   })
-  description = "The lock level to apply to the Key Vault. Possible values are `None`, `CanNotDelete`, and `ReadOnly`."
-  default     = {}
+  default     = {
+    kind = null
+  }
```
Comment (matt-FFFFFF) [inline docs/static/includes/interfaces/int.locks.schema.tf:8]:
```suggestion
  default     = {}
```

File: docs/static/includes/interfaces/int.locks.schema.tf [hunk 3]
```diff
@@ -1,21 +1,31 @@
...
+
+- `kind` - (Required) The type of lock. Possible values are `null`, `CanNotDelete` or `ReadOnly`.
+- `name` - (Optional) The name of the lock. If not specified a name will be generated based on the `kind` value. Changing this forces a new resource to be created.
+DESCRIPTION
   nullable    = false
+
   validation {
-    condition     = contains(["CanNotDelete", "ReadOnly", "None"], var.lock.kind)
-    error_message = "The lock level must be one of: 'None', 'CanNotDelete', or 'ReadOnly'."
+    condition     = var.lock.kind != null ? contains(["CanNotDelete", "ReadOnly"], var.lock.kind) : true
```
Comment (matt-FFFFFF) [inline docs/static/includes/interfaces/int.locks.schema.tf:18]:
```suggestion
    condition     = var.lock.kind == null || try(contains(["CanNotDelete", "ReadOnly"], var.lock.kind), false)
```

File: docs/static/includes/interfaces/int.locks.schema.tf [hunk 4]
```diff
@@ -1,21 +1,31 @@
...
+    error_message = "The lock level must be one of: `null`, `CanNotDelete` or `ReadOnly`."
```
Comment (matt-FFFFFF) [inline docs/static/includes/interfaces/int.locks.schema.tf:19]:
```suggestion
    error_message = "The lock level must be one of: `null`, `\"CanNotDelete\"` or `\"ReadOnly\"`."
```

File: docs/static/includes/interfaces/int.locks.schema.tf [hunk 5]
```diff
@@ -1,21 +1,31 @@
...
   })
-  description = "The lock level to apply to the Key Vault. Possible values are `None`, `CanNotDelete`, and `ReadOnly`."
-  default     = {}
+  default     = {
+    kind = null
+  }
+  description = <<DESCRIPTION
+Controls the Resource Lock configuration on this resource. The following properties can be specified:
+
+- `kind` - (Required) The type of lock. Possible values are `null`, `CanNotDelete` or `ReadOnly`.
```
Comment (matt-FFFFFF) [inline docs/static/includes/interfaces/int.locks.schema.tf:12]:
```suggestion
- `kind` - (Required) The type of lock. Possible values are `null`, `\"CanNotDelete\"` or `\"ReadOnly\"`.
```

**Review by matt-FFFFFF — CHANGES_REQUESTED** [review]

**Comment by MariusStorhaug** [comment]
Well in contrast to the IM interface, the lock interface needs the `kind` variable provided in order to deploy the lock. The `managed_identities` has two attributes which indeed are marked as optional as you can provide one or the other or both, but one of the two attributes are actually required for the module to configure the identity feature on the resource.

Will go ahead with your code suggestions if you still disagree with my suggestions above.

**Comment by matt-FFFFFF** [comment]
@MariusStorhaug Both approaches are semantically the same. Syntactically I prefer my proposal as it provides consistency with other interfaces and still fulfils the idiom of not specifying something means you don't get it.

**Comment by MariusStorhaug** [comment]
@matt-FFFFFF Understood, all suggestions added.

**Comment by matt-FFFFFF** [comment]
Hold fire. I've been thinking further on this and will discuss in the issue

**Comment by matt-FFFFFF** [comment]
@MariusStorhaug 

Apologies for the flip flopping here - I agree with your original suggestion but have simplified the validation rule slightly:

```terraform

Apologies for radio silence here and for the flip flopping.

I have tested this and am happy with the code:

variable "lock" {
  type = object({
    kind = string
    name = optional(string, null)
  })
  default = null
  validation {
    error_message = "Lock kind must be either `CanNotDelete` or `ReadOnly`"
    condition     = var.lock != null ? contains(["CanNotDelete", "ReadOnly"], var.lock.kind) : true
  }
}

resource "terraform_data" "test" {
  count = var.lock != null ? 1 : 0
  input = {
    kind = var.lock.kind
    name = coalesce(var.lock.name, "lock-${var.name}")
  }
}

```

**Comment by MariusStorhaug** [comment]
Update back to the original suggestion with a nullable lock variable and mandatory `kind` property, but merged the condition to be that of the other [suggested change, which was later reverted based on comments](https://github.com/Azure/Azure-Verified-Modules/pull/489#discussion_r1441825335).

**Review by matt-FFFFFF — APPROVED** [review]

**Comment by matt-FFFFFF** [comment]
Thank you @MariusStorhaug - sorry again for the confusion

**Pull request merged**